Blog - Using GitHub Actions to Run, Test, Build, and Deploy Docker Containers (#420)
earthly/website: blog/_posts/2023-06-15-github-actions-and-docker.md

---
title: "Using GitHub Actions to Run, Test, Build, and Deploy Docker Containers"
categories:
  - Tutorials
toc: true
author: James Olaogun

internal-links:
 - Github Actions
 - Deployment
 - Testing
 - Build
 - Docker Container
---

[GitHub Actions](https://github.com/features/actions) is a flexible tool that enables developers to automate a variety of processes, including developing, testing, and deploying, right from their GitHub repositories. The automation of [Docker](https://www.docker.com/) containers is no exception since GitHub Actions also enables developers to automate the process of developing containerized applications. As a result, developers can save time and focus on improving the overall quality of their software.

> In this article, you'll learn how to use [Github Actions](/blog/continuous-integration) to run, test, build, and deploy Docker containers using GitHub Actions.

Reviewer: ```suggestion
In this article, you'll learn how to use Github Actions to run, test, build, and deploy Docker containers using GitHub Actions.
```

## What Is GitHub Actions?

As previously stated, GitHub Actions is a platform for automating software workflows directly from a GitHub repository. With GitHub Actions, you can create workflows that are triggered by events, such as a push to the main branch or the creation of a new pull request in your GitHub repo. These workflows then run a series of actions, such as checking the code for syntax errors, building a Docker image, or deploying to a hosting platform.

In this tutorial, you'll be making use of the [GitHub Actions starter workflow](https://docs.github.com/en/actions/using-workflows/using-starter-workflows) examples. These examples can be used as a starting point for your automation pipelines and provide a good foundation for understanding how GitHub Actions works.

## How to Run, Test, Build, and Deploy Docker Containers Using GitHub Actions

Before you begin this tutorial, you'll need the following:

- [GitHub account](https://github.com)
- Basic knowledge of Docker and Docker Compose
- Basic knowledge of YAML files

> Once these prerequisites are met, it's time to begin. You'll start by creating a sample workflow, and then set a runner for the workflow. After that, you'll learn how to set up [GitHub](/blog/ci-comparison) Actions locally and then how to set up the build and test stage. Finally, you'll execute the workflow by running the action.

Reviewer: ```suggestion
Once these prerequisites are met, it's time to begin. You'll start by creating a sample workflow, and then set a runner for the workflow. After that, you'll learn how to set up GitHub Actions locally and then how to set up the build and test stage. Finally, you'll execute the workflow by running the action.
```

### Create a Workflow

> The first thing you need to do is create a workflow using GitHub Actions. This workflow defines the steps that GitHub Actions should take when triggered, including checking out the code, building a Docker [container](/blog/docker-slim), running tests, and deploying the application.

Reviewer: ```suggestion
The first thing you need to do is create a workflow using GitHub Actions. This workflow defines the steps that GitHub Actions should take when triggered, including checking out the code, building a Docker container, running tests, and deploying the application.
```

The `name` key can be `Build docker images`, the `id` key will have any unique string, and the `run` key will contain the build command as the value:

~~~{.yml caption=""}
- name: Build docker images
  Id: build-image
  run: | 
    echo ---Building images and starting up docker---
    {{docker build [image-url] or docker-compose -f \
    [docker-compose file] up -d }}
    echo ---Containers up—
~~~

Following the **Deploy to Amazon ECS** workflow template that's been used here, the build step can be found on line 67:

~~~{.yml caption=""}
- name: Build, tag, and push image to Amazon ECR
      id: build-image
      env:
        ECR_REGISTRY: ${{ steps.login-ecr.outputs.registry }}
        IMAGE_TAG: ${{ github.sha }}
      run: |
        # Build a docker container and
        # push it to ECR so that it can
        # be deployed to ECS.
        docker build -t $ECR_REGISTRY/$ECR_REPOSITORY:$IMAGE_TAG .
        docker push $ECR_REGISTRY/$ECR_REPOSITORY:$IMAGE_TAG
        echo "image=$ECR_REGISTRY/$ECR_REPOSITORY:$IMAGE_TAG" >> \
        $GITHUB_OUTPUT
~~~

This template's build step uses the `env` key since `ECR_REGISTRY` requires a login and SHA key.

### Set Up the Test Stage

> Once you're done setting up the build stage, the next thing you need to do is set up the test stage. Because the **Deploy to Amazon ECS** workflow template is primarily a [deployment](/blog/deployment-strategies) template, the test stage isn't factored into it. However, the test stage is the recommended phase since it helps you verify the functionality and connections of your applications (including the workflow file) before pushing to GitHub to run the workflow actions.

Reviewer: ```suggestion
Once you're done setting up the build stage, the next thing you need to do is set up the test stage. Because the **Deploy to Amazon ECS** workflow template is primarily a deployment template, the test stage isn't factored into it. However, the test stage is the recommended phase since it helps you verify the functionality and connections of your applications (including the workflow file) before pushing to GitHub to run the workflow actions.
```

To set up the test stage, you can either create a test folder in your application or in a different folder outside your application, then develop all your test cases. Afterward, ensure your test folder is containerized in the same network as your application. If you're using [Docker Compose](/blog/youre-using-docker-compose-wrong), GitHub Actions will run the workflow in a user-defined network. And for communication to happen, your application and test folder must be in the same network. The network in this context is the Docker Compose network.

> The purpose of this network is to provide a Docker network for the containers defined in the `docker-compose` file, allowing them to communicate with each other and with the host system. Containers in separate networks cannot communicate, which means that API test cases won't be able to get a response from the services in other networks. You can learn more about creating a `docker-compose` file with a network [in this article](https://medium.com/@caysever/docker-compose-network-b86e424fad82).

Reviewer: ```suggestion
The purpose of this network is to provide a Docker network for the containers defined in the `docker-compose` file, allowing them to communicate with each other and with the host system. Containers in separate networks cannot communicate, which means that API test cases won't be able to get a response from the services in other networks. You can learn more about creating a `docker-compose` file with a network [in this article](/blog/docker-networking).
```

> When you're done adding the test cases folder to the same [container](/blog/docker-slim) network, you need to add a new value to the `steps` key of your workflow file. The new value will contain a `name`, `id`, and `run` key:

Reviewer: ```suggestion
When you're done adding the test cases folder to the same container network, you need to add a new value to the `steps` key of your workflow file. The new value will contain a `name`, `id`, and `run` key:
```

In addition, Earthly allows you to define reusable builds that you can use across different machines, making it easier to collaborate and enabling you to run your builds anywhere.

Moreover, Earthly caches your build components, making the process efficient and fast. Check out the [official documentation](https://docs.earthly.dev/basics) to learn more about getting started with Earthly.
</div>

### Run the Action

After you've set up the test stage, the final step is to run the action. To do that, you need to push the changes made to your workflow file to your GitHub repository. The workflow will be triggered based on the branches you set it to monitor on the `on` key of your workflow file.

You can also monitor the progress of the workflow by visiting the **Actions** tab in your repository on GitHub and selecting a workflow that has previously been run or the currently running workflow:

<div class="wide">
![Workflow history]({{site.images}}{{page.slug}}/VQCqOYR.png)
</div>

## Conclusion

> In this tutorial, you learned how to create a new workflow, edit an existing workflow, set up the runner for your workflow, and locally set up and work with GitHub Actions. At this point, you should be confident that you can build [GitHub](/blog/ci-comparison) Actions for your projects and speed up the development processes.

Reviewer: ```suggestion
In this tutorial, you learned how to create a new workflow, edit an existing workflow, set up the runner for your workflow, and locally set up and work with GitHub Actions. At this point, you should be confident that you can build Github Actions for your projects and speed up the development processes.
```

After reading this article, it's recommended that you explore and experiment with the various actions, events, and workflows available in the GitHub Actions ecosystem to further enhance and automate your development processes.

> 

Reviewer: ```suggestion
If you want to see how Github Actions stacks up against other CI services, check out our comparison of [CI Free Tiers](/blog/ci-comparison).
```
